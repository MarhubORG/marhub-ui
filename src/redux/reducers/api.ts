import { ApiState } from '../../types/interfaces';
import {
  EXPORTING_IRAP_DATA,
  EXPORT_IRAP_DATA_FAILURE,
  EXPORT_IRAP_DATA_SUCCESS,
  LOGOUT,
} from '../constants/actionTypes';
import { ApiActionTypes } from '../actions/api';

export const initialState = {
  loading: false,
  irapState: {},
  error: '',
};

export default function apiReducer(
  state = initialState,
  action: ApiActionTypes
): ApiState {
  switch (action.type) {
    case EXPORTING_IRAP_DATA:
      return { ...state, loading: true };
    case EXPORT_IRAP_DATA_SUCCESS:
      return { ...state, loading: false, irapState: action.payload };
    case EXPORT_IRAP_DATA_FAILURE:
      return { ...state, loading: false, error: action.payload };
    case LOGOUT:
      return { ...initialState };
    default:
      return state;
  }
}
