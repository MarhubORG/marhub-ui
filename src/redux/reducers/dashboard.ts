import { DashboardState } from '../../types/interfaces';
import {
  FETCH_ORGANIZATIONS,
  FETCH_ORGANIZATIONS_FAILURE,
  FETCH_ORGANIZATIONS_SUCCESS,
  UPDATE_ORGANIZATION,
  UPDATE_ORGANIZATION_FAILURE,
  UPDATE_ORGANIZATION_SUCCESS,
  LOGOUT,
  CREATE_ORGANIZATION,
  CREATE_ORGANIZATION_FAILURE,
  CREATE_ORGANIZATION_SUCCESS,
  CREATE_ORGANIZATION_REDIRECT,
  CREATE_TEMPLATE,
  CREATE_TEMPLATE_FAILURE,
  CREATE_TEMPLATE_SUCCESS,
  UPDATE_TEMPLATE,
  UPDATE_TEMPLATE_FAILURE,
  UPDATE_TEMPLATE_SUCCESS,
  DELETE_TEMPLATE,
  DELETE_TEMPLATE_FAILURE,
  DELETE_TEMPLATE_SUCCESS,
  DELETE_TEMPLATE_REDIRECT,
} from '../constants/actionTypes';
import { DashboardActionTypes, Organization } from '../actions/dashboard';

export const standardFetchOrganizationErrorMessage =
  'Could not fetch organizations. Please try again later or contact the administrator.';
export const templateFailureMessage =
  'Creating template failed. Please contact your administrator.';
export const templateSuccessMessage = 'Successfully created new template';

export const initialState = {
  loading: false,
  organizations: [],
  errorMessage: '',
  redirectToVisibleFields: '',
  templateMessage: '',
};
export const successMessage = 'Successfully updated organization permissions.';
export const createOrgFailureMessage =
  'Creating a new organization failed. Please contact your administrator.';
export const createOrgSuccessMessage =
  'Successfully updated organization permissions.';
export default function dashboardReducer(
  state = initialState,
  action: DashboardActionTypes
): DashboardState {
  switch (action.type) {
    case FETCH_ORGANIZATIONS:
      return { ...state, loading: true, errorMessage: '' };
    case FETCH_ORGANIZATIONS_SUCCESS:
      return {
        ...state,
        loading: false,
        organizations: [...action.payload],
        errorMessage: '',
      };
    case FETCH_ORGANIZATIONS_FAILURE:
      return {
        ...state,
        loading: false,
        errorMessage: standardFetchOrganizationErrorMessage,
      };
    case UPDATE_ORGANIZATION:
      return { ...state, loading: true, errorMessage: '' };
    case UPDATE_ORGANIZATION_FAILURE:
      return { ...state, loading: false, errorMessage: action.payload };
    case UPDATE_ORGANIZATION_SUCCESS:
      return {
        ...state,
        loading: false,
        errorMessage: successMessage,
        organizations: replaceWithUpdatedOrg(action.payload, state),
      };
    case LOGOUT:
      return { ...initialState };
    case CREATE_ORGANIZATION:
      return {
        ...state,
        loading: true,
        errorMessage: '',
      };
    case CREATE_ORGANIZATION_FAILURE:
      return {
        ...state,
        loading: false,
        errorMessage: createOrgFailureMessage,
      };
    case CREATE_ORGANIZATION_SUCCESS:
      return {
        ...state,
        organizations: [...state.organizations, action.payload],
        loading: false,
        errorMessage: createOrgSuccessMessage,
        redirectToVisibleFields: action.payload.organisation.name,
      };
    case CREATE_ORGANIZATION_REDIRECT:
      return {
        ...state,
        redirectToVisibleFields: '',
      };
    case CREATE_TEMPLATE_FAILURE:
      return {
        ...state,
        templateMessage: templateFailureMessage,
        loading: false,
      };
    case CREATE_TEMPLATE:
      return {
        ...state,
        loading: true,
        templateMessage: '',
      };
    case CREATE_TEMPLATE_SUCCESS:
      return {
        ...state,
        loading: false,
        templateMessage: templateSuccessMessage,
        organizations: replaceWithUpdatedOrg(action.payload, state),
      };
    case UPDATE_TEMPLATE:
      return {
        ...state,
        loading: true,
        templateMessage: '',
      };
    case UPDATE_TEMPLATE_FAILURE:
      return {
        ...state,
        loading: false,
        templateMessage: action.payload,
      };
    case UPDATE_TEMPLATE_SUCCESS:
      return {
        ...state,
        loading: false,
        templateMessage: 'Updated template successfully',
        organizations: replaceWithUpdatedOrg(action.payload, state),
      };
    case DELETE_TEMPLATE:
      return {
        ...state,
        loading: true,
        templateMessage: '',
      };
    case DELETE_TEMPLATE_FAILURE:
      return {
        ...state,
        loading: false,
        templateMessage: action.payload,
      };
    case DELETE_TEMPLATE_SUCCESS:
      return {
        ...state,
        loading: false,
        templateMessage: 'Template Deleted.',
        organizations: replaceWithUpdatedOrg(action.payload, state),
      };
    case DELETE_TEMPLATE_REDIRECT:
      return {
        ...state,
        loading: false,
        templateMessage: '',
      };
    default:
      return state;
  }
}

export function replaceWithUpdatedOrg(
  organization: Organization,
  state: DashboardState
): Organization[] {
  return state.organizations.map(el => {
    if (el.id === organization.id) return organization;
    return el;
  });
}
