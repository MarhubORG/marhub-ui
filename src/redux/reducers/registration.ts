import { RegistrationState } from '../../types/interfaces';
import {
  SIGNUP,
  SIGNUP_SUCCESS,
  SIGNUP_ERROR,
  SIGNUP_REDIRECTING,
  LOGIN,
  LOGIN_ERROR,
  LOGIN_SUCCESS,
  LOGIN_REDIRECTING,
} from '../constants/actionTypes';
import { RegistrationActionTypes } from '../actions/index';

export function registrationReducer(
  state = initialState,
  action: RegistrationActionTypes
): RegistrationState {
  switch (action.type) {
    case SIGNUP:
      return { ...state, loading: true };
    case SIGNUP_SUCCESS:
      return { ...state, loading: false, redirect: true };
    case SIGNUP_REDIRECTING:
      return { ...state, redirect: false };
    case SIGNUP_ERROR:
      return {
        ...state,
        loading: false,
        isLoggedIn: false,
        error: action.payload.message,
      };
    case LOGIN:
      return { ...state, loading: true };
    case LOGIN_ERROR:
      return {
        ...state,
        loading: false,
        isLoggedIn: false,
        error: action.payload.message,
      };
    case LOGIN_SUCCESS:
      return {
        ...state,
        loading: false,
        loginRedirect: true,
        isLoggedIn: true,
        role: action.payload.role,
      };
    case LOGIN_REDIRECTING:
      return { ...state, loginRedirect: false };
    default:
      return state;
  }
}

export default registrationReducer;

export const initialState: RegistrationState = {
  loading: false,
  isLoggedIn: false,
  redirect: false,
  loginRedirect: false,
  error: '',
  role: '',
};
