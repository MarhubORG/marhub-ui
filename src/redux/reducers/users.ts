import { UserState } from '../../types/interfaces';
import { UserActionTypes } from '../actions/users';
import {
  FETCH_USERS,
  FETCH_USERS_FAILURE,
  FETCH_USERS_SUCCESS,
  LOGOUT,
} from '../constants/actionTypes';

const failureMessage =
  'Loading Users failed. Please contact your administrator.';
const initialState = {
  users: [],
  message: '',
  loading: false,
};
export default function userReducer(
  state = initialState,
  action: UserActionTypes
): UserState {
  switch (action.type) {
    case FETCH_USERS:
      return { ...state, loading: true };
    case FETCH_USERS_FAILURE:
      return { ...state, loading: false, message: failureMessage };
    case FETCH_USERS_SUCCESS:
      return { ...state, loading: false, message: '', users: action.payload };
    case LOGOUT:
      return initialState;
    default:
      return state;
  }
}
