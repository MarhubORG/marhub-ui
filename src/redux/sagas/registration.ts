import axios from 'axios';
import { put, takeLatest } from 'redux-saga/effects';
import cookie from 'react-cookies';

import { SIGNUP, LOGIN } from '../constants/actionTypes';
import {
  signupSuccess,
  signupError,
  loginError,
  loginSuccess,
  SignupAction,
  LoginAction,
} from '../actions/index';

function* signup(action: SignupAction) {
  try {
    const url = 'http://localhost:8080/api/v1/register';
    yield axios.post(url, {
      name: action.payload.name,
      organisation: action.payload.organization,
      email: action.payload.email,
      password: action.payload.password,
    });
    yield put(signupSuccess());
  } catch (error) {
    yield put(signupError(error.response.data.message));
  }
}

export function* actionWatcher() {
  yield takeLatest(SIGNUP, signup);
}

function* login(action: LoginAction) {
  try {
    const url = 'http://localhost:8080/api/v1/login';
    console.log({ url });
    const json = yield axios.post(url, {
      email: action.payload.email,
      password: action.payload.password,
    });
    console.log({ json });
    cookie.save('token', json.data.user.sessionToken, {
      path: '/',
      // domain: 'http://localhost:8080',
      // secure: true,
      // httpOnly: true,
    });
    yield put(loginSuccess());
  } catch (error) {
    yield put(loginError(error.response.data.message));
  }
}

export function* watchLogin() {
  yield takeLatest(LOGIN, login);
}
