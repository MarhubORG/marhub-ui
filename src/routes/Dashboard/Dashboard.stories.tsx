import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { ThemeProvider } from 'styled-components';
import { mainTheme } from '../../styles/themes';
import { UnconnectedDashboard } from './Dashboard';

storiesOf('Dashboard', module).add('Dashboard Component when logged in', () => (
  <ThemeProvider theme={mainTheme}>
    <UnconnectedDashboard
      myOrganization="myOrganization"
      role="role"
      isLoggedIn
    />
  </ThemeProvider>
));
