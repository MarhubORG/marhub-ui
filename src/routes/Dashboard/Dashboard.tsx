import React from 'react';
import styled from 'styled-components';
import cookie from 'react-cookies';
import { RouteComponentProps } from '@reach/router';
import { connect } from 'react-redux';
import { RootState } from '../../types/interfaces';
import Login from '../Registration/Login/Login';

interface DashboardProps extends RouteComponentProps {
  isLoggedIn: boolean;
}

export function UnconnectedDashboard(props: DashboardProps): JSX.Element {
  if (props.isLoggedIn !== true) {
    return <Login />;
  }
  return (
    <div>
      <Button>Download IRAP Data</Button>
    </div>
  );
}

export interface MapStateToProps {
  isLoggedIn: boolean;
}

export function mapStateToProps(state: RootState): MapStateToProps {
  const { isLoggedIn } = state.registration;
  return { isLoggedIn };
}

const Button = styled.button`
  margin: 1rem;
  width: 12rem;
  height: 3rem;
  background-color: ${({ theme }): string => theme.primaryColor};
  color: white;
  font-family: Open Sans, sans-serif;
  font-weight: 700;
  border-radius: 0.2rem;
  font-size: 0.8rem;
`;

export default connect(mapStateToProps)(UnconnectedDashboard);
