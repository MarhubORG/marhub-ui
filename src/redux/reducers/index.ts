import { combineReducers } from 'redux';
import registration from './registration';
import apiReducer from './api';
import dashboardReducer from './dashboard';

export default combineReducers({
  registration,
  apiReducer,
  dashboardReducer,
});
