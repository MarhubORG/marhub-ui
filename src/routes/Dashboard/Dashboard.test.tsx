import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import { UnconnectedDashboard, mapStateToProps } from './Dashboard';
import { initialState } from '../../redux/reducers/registration';

const mockStore = configureMockStore();
const store = mockStore({ registration: initialState });

it('renders correctly when logged in', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <UnconnectedDashboard isLoggedIn />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly when not logged in', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <UnconnectedDashboard isLoggedIn />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

describe('mapStateToProps', () => {
  it('contains registration state', () => {
    expect(mapStateToProps({ registration: initialState })).toBeDefined();
    expect(mapStateToProps({ registration: initialState }).isLoggedIn).toEqual(
      false
    );
  });
});
