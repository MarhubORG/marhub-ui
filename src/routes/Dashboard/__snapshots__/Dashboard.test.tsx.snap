// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when logged in 1`] = `
.c0 {
  margin: 1rem;
  width: 12rem;
  height: 3rem;
  color: white;
  font-family: Open Sans,sans-serif;
  font-weight: 700;
  border-radius: 0.2rem;
  font-size: 0.8rem;
}

<div>
  <button
    className="c0"
  >
    Download IRAP Data
  </button>
</div>
`;

exports[`renders correctly when not logged in 1`] = `
.c0 {
  margin: 1rem;
  width: 12rem;
  height: 3rem;
  color: white;
  font-family: Open Sans,sans-serif;
  font-weight: 700;
  border-radius: 0.2rem;
  font-size: 0.8rem;
}

<div>
  <button
    className="c0"
  >
    Download IRAP Data
  </button>
</div>
`;
